src/map.h

```diff
@@ -374,14 +378,43 @@ class ServerMap : public Map
 	void beginSave();
 	void endSave();
 
-	void save(ModifiedState save_level);
+	/*
+	 * For async save these methods must be used as follows:
+	 *
+	 * SerializedBlockList list;
+	 * CompressedBlockList result;
+	 * lock(env mutex);
+	 *   map.save(..., true);
+	 *   map.timerUpdate(...); // optional
+	 *   swapSerializedBlocks(list)
+	 * unlock(env mutex);
+	 *
+	 * map.compress(list, result);
+	 *
+	 * lock(env mutex);
+	 *   map.save(result);
+	 * unlock(env mutex);
+	 *
+	 * save(..., false) performs all these step,
+	 * but has to be under the env lock completely
+	 */
+	void save(ModifiedState save_level, bool async=false);
+	void save(CompressedBlockList &blocks);
+	void swapSerializedBlocks(SerializedBlockList &new_list) {
+		swap(m_serialized_blocks, new_list);
+	}
```

Comment:
How about something like this? While not technically correct, I think this is easier to understand:
```C++
// Transfers ownership of the serialized blocks to the passed list
SerializedBlockList &getSerializedBlocks(SerializedBlockList &list)
{
	assert(list.empty());
	std::swap(m_serialized_blocks, list);
}
```

A desired approach would be to `std::swap` it to a local variable and return that directly. Sadly that relies on NRVO which isn't guaranteed.

Reply:
Yep. That's fine. I was struggling with how to name this method.

Another approach I explored was writing these objects to a queue and then the save-thread popping objects off the queue. That would have needed extra locking.